Hide headers if webmentions are not supported to that URL

dshanske opened the pull request:
Address #75 and #152 

The problem with hiding headers prior was partially that is_singular() or anything requiring a global query variable isn't loaded when headers are sent, so it would always evaluate to false.

This uses get_self_link(scheduled for WordPress 5.3, see https://core.trac.wordpress.org/ticket/44838) to retrieve the URL currently being requested, then determines the post_id if applicable using webmention_to_postid, which means the existing filter could take archive or other pages and redirect them. It also works for the currently built in home page mention redirect.

pfefferle commented:
What about homepage webmentions?

pfefferle commented:
In the WebSub plugin I use the `template_redirect` hook, which supports the `is_*` functions https://github.com/pubsubhubbub/wordpress-pubsubhubbub/blob/master/includes/class-pubsubhubbub-topics.php#L48

dshanske commented:
homepage mentions are covered because if not set, webmention url to post id will return 0

pfefferle commented:
But then it returns false and adds no header!?!

pfefferle commented:
What I mean is, it shows no headers on the frontage!

pfefferle commented:
Have you had a look at the websub code?

pfefferle commented:
And what about static front pages?

pfefferle commented:
And what about custom pages/archive pages https://github.com/pfefferle/wordpress-webmention#how-can-i-handle-webmentions-to-my-homepage-or-archive-pages

pfefferle commented:
This wouldn‘t work any more...

dshanske commented:
It would. I checked it. Because it would rely on the page it was redirecting to having pings open

dshanske commented:
So, the filter redirects archive/home pages to a post id. The function checks that post id for whether it accepts pings. If there is no post id, it. would return false

pfefferle commented:
Thanks for the clarification on the #indieweb chat! Looks promising!

pfefferle commented:
Perhaps we can move the pings_open check to the `webmention_url_to_postid` function?

dshanske commented:
That would retrieve the entire post every time we retrieved the page... although that technically would happen anyway.

dshanske commented:
I moved it, as suggested to the template_redirect hook as otherwise it would generate a WP_Query to find the post on every page load, and that could get expensive. Also added caching for when it is an archive page so that it would only run the query once per page load, as opposed to twice.

dshanske commented:
Also added the constant to disable it.

pfefferle commented on includes/class-webmention-receiver.php:
```diff
@@ -824,10 +824,43 @@ public static function comment_template_include( $template ) {
+		}
+		if ( ! is_singular() ) {
+			// Cache this as will be called twice in a page load
+			$post_id = wp_cache_get( base64_encode( get_self_link() ), 'wmurl' );
```
But this will only work if you have installed a cache plugin: https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Caching

dshanske replied:
Must disagree. That is if you want it to persist for multiple page loads. I want it to persist during the same page load...once for the http headers and once for the html headers

pfefferle replied:
Ah I see: https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Examples

dshanske replied:
If someone enables a persistent cache, then they'd derive an extra benefit

pfefferle replied:
I learned a lot today! thanks!

pfefferle commented:
thanks a lot @dshanske
